### PR TITLE
gBoards GergoPlex

### PR DESCRIPTION
**Pull request by pirj**
## Description

Add firmware for [gBoards GergoPlex](https://www.gboards.ca/product/gergoplex).

## Not done

 - `SPLIT_KEYBOARD=yes` could potentially reduce the amount of custom code
 - in the custom matrix code rows and cols seem to be mixed up. Might be it's intentional
 - the code is reading directly using `PINF` (inherited from #8928)

## Types of Changes

- [x] Keyboard (addition or update)

## Issues Fixed or Closed by This PR

fixes #8262, #8928, #10406

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by btannous** [comment]
looks like everything works except the mouse functionality, neither clicking or moving seem to work (which you mentioned the lmb and rmb do not seem to work)

File: keyboards/gboards/gergoplex/keymaps/default/rules.mk
Comment (btannous) [inline keyboards/gboards/gergoplex/keymaps/default/rules.mk:6]: This line to `yes` might fix the mouse issues.

**Comment by btannous** [comment]
https://github.com/pirj/qmk_firmware/pull/1

updated the delay:
 - it was set to 40
 - docs seem to show 300
 - so went with 400
 - https://beta.docs.qmk.fm/using-qmk/software-features/feature_combo

also the `mousekey_enable = no` in the keymap seemed to override whats expected, so changed that to yes.

seems much more usable with 400 vs 40!

File: keyboards/gboards/gergoplex/keymaps/default/rules.mk
Comment (pirj) [reply keyboards/gboards/gergoplex/keymaps/default/rules.mk:6]: Oh 🙈 
Thank you for spotting this!

**Comment by pirj** [comment]
@btannous Incorporated your fixes with an adjustment - removed the redundant `MOUSEKEY_ENABLE` from `keymaps/default/rules.mk` so it falls back to `rules.mk`; removed the `COMBO_TERM` so it falls back to `TAPPING_TERM` that is 200 by default.
Thank you!

**Comment by btannous** [comment]
@pirj awesome work.  `IGNORE_MOD_TAP_INTERRUPT` seems to fix my issue with the CTRL/Shift modifiers, so sent you a PR for that one line.

I'm using your latest branch + that config change and so far this seems to work well.

**Comment by pirj** [comment]
@btannous Thanks, merged!
I am sorry for thoughtlessly removing this option.

**Comment by btannous** [comment]
All good, that change makes a huge difference on those keys.  Now Im hopeful this is good to the QMK maintainers!  Thanks all :)

**Comment by btannous** [comment]
Also +@germ

File: keyboards/gboards/gergoplex/gergoplex.c
Comment (drashna) [inline keyboards/gboards/gergoplex/gergoplex.c:1]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

File: keyboards/gboards/gergoplex/gergoplex.c
Comment (drashna) [inline keyboards/gboards/gergoplex/gergoplex.c:8]:
```suggestion
```

File: keyboards/gboards/gergoplex/gergoplex.h
Comment (drashna) [inline keyboards/gboards/gergoplex/gergoplex.h:1]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

File: keyboards/gboards/gergoplex/keymaps/default/keymap.c
Comment (drashna) [inline keyboards/gboards/gergoplex/keymaps/default/keymap.c:6]:
```suggestion
#include QMK_KEYBOARD_H
```

File: keyboards/gboards/gergoplex/keymaps/default/keymap.c
Comment (drashna) [inline keyboards/gboards/gergoplex/keymaps/default/keymap.c:1]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

File: keyboards/gboards/gergoplex/matrix.c
Comment (drashna) [inline keyboards/gboards/gergoplex/matrix.c:19]: This should be converted to matrix lite.

Check the ergodox ez's matrix for reference.

File: keyboards/gboards/gergoplex/rules.mk
Comment (drashna) [inline keyboards/gboards/gergoplex/rules.mk:5]:
```suggestion
CUSTOM_MATRIX    = lite
```

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/gboards/gergoplex/matrix.c
Comment (pirj) [reply keyboards/gboards/gergoplex/matrix.c:19]: Done so in the last commit, but the keyboard just stopped working.
Can you please give a cue what did I do wrong, @drashna ?

Comment (drashna) [reply keyboards/gboards/gergoplex/matrix.c:19]: Hmm.  My guess is that some of the config isn't correct. 

It's just the expander side that stopped working?

Comment (btannous) [reply keyboards/gboards/gergoplex/matrix.c:19]: In `static matrix_row_t read_cols(uint8_t row) {` @germ originally did a write before reading i2c, it looks like that's missing, maybe thats it? (I'm new to qmk_firmware, so not sure)

https://github.com/germ/qmk_firmware/blob/39c86e080dc04b68ee08387dcb5144c88cb44855/keyboards/gboards/k/gergoplex/matrix.c#L223-L224

Comment (pirj) [reply keyboards/gboards/gergoplex/matrix.c:19]: > It's just the expander side that stopped working?

Both, actually. Is this a clue?

Comment (drashna) [reply keyboards/gboards/gergoplex/matrix.c:19]: I'm sorry, I lost track of this. 

And yeah, it means that the matrix isn't being processed correctly.  Give me a couple of days, and I'll see if I can fix up some code. 

Unfortunately, I don't have a gergoplex to test on, and I'm not sure how close the other boards are in design, so testing is ... problematic.  

If it's more than a few days, give me a ping on the discord.

Comment (pirj) [reply keyboards/gboards/gergoplex/matrix.c:19]: Thanks. I might have some time this weekend to debug it. On the one hand (pun intended) it's confusing that both parts of the keyboard stop working, even though one of the matrices is scanned via I2C, and another with matrix lite's helper code.
The other thing that confuses me is that rows and columns are mixed up.
A good thing is that if you checkout `@~1`, you have a working code.

Comment (drashna) [reply keyboards/gboards/gergoplex/matrix.c:19]: Yeah, that design is pulled from the ergodox (og), and ... IIRC it's to help with scan speeds

**Comment by George-Lee** [comment]
Is there anything I can do as an amateur with C to help with this? (waiting on my gergoplex to be shipped so no device yet)

**Comment by stjoni** [comment]
I have not attempted to debug, but merging manna-harbour/miryoku into this branch and `make gboards/gergoplex:manna-harbour_miryoku` builds without errors, but the keyboard is not functional after flashing.

**Comment by caksoylar** [comment]
@stjoni I believe the latest commit in the PR is still WIP with the custom matrix lite refactor. Commits up to and including https://github.com/qmk/qmk_firmware/pull/13027/commits/06b85ec0376fe5598ac98ff271167a484044208b builds and works fine on my end.

**Comment by drashna** [comment]
For now, it may be simplest/best to revert the matrix lite code.

**Comment by pirj** [comment]
@drashna Thank you! Will handle it over the weekend.
It was a slight improvement code-wise, but not as significant as I imagined it would be to reuse the commonly used code.

**Comment by pirj** [comment]
@drashna I've reset the commit with matrix lite.

File: keyboards/gboards/gergoplex/rules.mk
Comment (caksoylar) [inline keyboards/gboards/gergoplex/rules.mk:16]:
```suggestion
DEBOUNCE_TYPE  = sym_eager_pr
```
This fails building with current `develop` since the old names seem to be removed.

Comment (pirj) [reply keyboards/gboards/gergoplex/rules.mk:16]: Thanks!

**Comment by pirj** [comment]
Rebased on `develop`.

**Comment by pirj** [comment]
Something's not right. I've reinstalled QMK package in my system, and rebased the PR on `develop`. Now the previously working (pre matrix lite) firmware doesn't work. It builds, flashes, but the keyboard does not react.
I've checked out my `prettify-gergoplex-with-matrix-lite` that isn't rebased onto recent `develop`, checked out `HEAD~1` to remove matrix lite changes, the same issue.
I've checked out Germ's `gboards-gergoplex` (#8928), but it doesn't even build.

My suspicion is that it's me removing duplicated method definitions due to the newly appeared linker error:
```
quantum/matrix_common.c:36: multiple definition of `matrix_is_on'; .build/obj_gboards_gergoplex_default/matrix.o:/Users/pirj/source/qmk_firmware/keyboards/gboards/gergoplex/matrix.c:160: first defined here
```
I had to remove those methods for the build to work.

However, a warning has caught my eye yesterday, something like `MATRIX_ROWS doesn't match MATRIX_ROWS_PINS, 10 vs 4` and a similar related to columns.
Keeping in mind that rows and cols are horribly mixed up here:
```
// Remember this means ROWS
static void init_cols(void) {
    for (uint8_t row = 0; row < MATRIX_COLS; row++) {
```
```
static const pin_t row_pins[MATRIX_COLS] = MATRIX_ROW_PINS;
// Right-hand side only pins, the left side is controlled my MCP
static const pin_t col_pins[MATRIX_ROWS_PER_SIDE] = MATRIX_COL_PINS;
```
I suspect that the definition of `matrix_key_count` from `quantum/matrix.c`:
```
uint8_t matrix_key_count(void) {
    uint8_t count = 0;
    for (uint8_t i = 0; i < MATRIX_ROWS; i++) {
        count += matrix_bitpop(i);
    }
    return count;
}

#if (MATRIX_COLS <= 8)
#    define matrix_bitpop(row) bitpop(matrix_get_row(row))
#elif (MATRIX_COLS <= 16)
#    define matrix_bitpop(row) bitpop16(matrix_get_row(row))
```

doesn't do good job replacing our definition (`keyboards/gboards/gergoplex/matrix.c`) when provided an incorrect `MATRIX_COLS`:
```
uint8_t matrix_key_count(void) {
    uint8_t count = 0;
    for (uint8_t i = 0; i < MATRIX_ROWS; i++) {
        count += bitpop16(matrix[i]);
    }
    return count;
}
```

Thus is due to `MATRIX_COLS` being set to 4 instead of 10. And, as a result, `bitpop` used instead of `bitpop16`.

I tried to confirm this theory by temporarily patching `quantum/matrix_common.c`:
```diff
#if (MATRIX_COLS <= 8)
- #    define matrix_bitpop(row) bitpop(matrix_get_row(row))
+ #    define matrix_bitpop(row) bitpop16(matrix_get_row(row))
```

but the firmware still didn't work.

The last working firmware is at https://github.com/btannous/qmk_firmware/tree/prettify-gergoplex

I must have broken something after that.

**Comment by pirj** [comment]
Silly me, I forgot to remove two commits related to custom matrix lite. Removed. Now everything works fine. 🎉 

@btannous I've incorporated your [changes for Colemak Mod-DH](https://github.com/pirj/qmk_firmware/pull/3). I've kept `germ-mouse-keys.def` separate, just in case someone would want to exclude them, and it's as easy as editing their `combos.def`. Thanks!

File: keyboards/gboards/gergoplex/keymaps/colemak-dhm/rules.mk
Comment (manna-harbour) [inline keyboards/gboards/gergoplex/keymaps/colemak-dhm/rules.mk:17]: Enable [community layout](https://beta.docs.qmk.fm/developing-qmk/qmk-reference/feature_layouts) support:

```suggestion
endif

LAYOUTS = split_3x5_3
```

**Comment by George-Lee** [comment]
Is there anything we/I can do to help with this?

**Review by drashna — APPROVED** [review]

**Comment by btannous** [comment]
YAY

**Comment by btannous** [comment]
Typing this on my georgoplex after I checked out https://github.com/pirj/qmk_firmware/tree/prettify-gergoplex and merged with a freshly synced master then flashed with qmk on an m1 chip. all works well. cant wait till this finally is going to get merged in.  thanks everyone!

**Review by btannous — APPROVED** [review]

**Comment by pirj** [comment]
Thank you, **@germ**, @btannous, @manna-harbour, **dear maintainers**, and early testers @YorickPeterse, @stjoni and @caksoylar for this to happen! 🎉

**Pull request merged**
